modules/zlib/metadata.json

```diff
@@ -7,7 +7,9 @@
         }
     ],
     "versions": [
-        "1.2.11"
```

Comment:
Removing `1.2.11` will cause other modules that depend on this version to break, we should not do this.

Reply:
I'm conflicted here about agreeing with you. From a reproducibility standpoint, definitely, we should not remove existing versions of a package. However in this case, the advice from upstream is that 1.2.12 is a drop in replacement for 1.2.11 that fixes a number of security issues and that 1.2.11 should no longer be used. So with my security hat on I would not agree with this advice and would expect that a dependency on 1.2.11 should in fact resolve to 1.2.12 as this is the minimum version that is 'secure' and that as the module maintainer I have baked in this knowledge into the registry by removing the 1.2.11 package. I think the fact that the end user has bumped the registry commit to use a HEAD that includes this change should mean that they have consciously updated and so should expect that package resolution could have changed, including updates to packages for security fixes. I'm assuming that the current implementation of min version selection in bazel doesn't do this kind of resolution of selecting the minimum version that exists in the registry?
On the bazel slack I suggested this as a way of solving the security type of use case:

> one way a yanking like mechanism could be implemented is to introduce a concept like python uses of having a constraints dependency; you could imagine a `bzl_constraint("zlib", min_version="1.2.11")` type construct that sets a minimum version. You could then have a security policy module consisting only of bzl_constraint that packages can depend on so you can globally set min versions of packages. You could also have a `bzl_constraint("zlib", exclude_version="1.2.11")` option that would turn off just one version of a particular package.

but I think that the logic in bzlmod should also be extended to do the kind of min version selection I suggested above i.e. a dep on version 1.2.11 should resolve to 1.2.12 or later if a specific version 1.2.11 does not exist in the registry without requiring a module to explicitly depend on 1.2.12 (otherwise a security update needs to review all packages that depend on itself and explicitly update those requirements without a constraint like mechanism I described above.

Reply:
And just to be clear here - I deliberately made this change of removing 1.2.11 from versions to invoke this discussion as I don't think the design space has been properly explored yet for deprecating / yanking versions (and the yanked_versions design may even be incorrect/incomplete as a result of this?).

Reply:
Thanks for invoking this discussion! I actually had the same hesitation and I agree this problem hasn't been studied throughly. I think this deserves a thread in https://groups.google.com/a/bazel.build/g/external-deps or a short design doc. I can do it and share my thoughts when I have more free time (probably next week), but feel free to start the conversation with the community first.

> `bzl_constraint("zlib", min_version="1.2.11")`

This is actually `bazel_dep` but without giving visibility to this dependency. It definitely could be (part of) the solution.

Reply:
Sorry for the delay here, I finally had time to start a discussion at https://github.com/bazelbuild/bazel-central-registry/discussions/128